packages/xo-web/src/common/modal.js

```diff
@@ -261,57 +261,127 @@ export const confirm = ({ body, icon = 'alarm', title, strongConfirm }) =>
 
 // -----------------------------------------------------------------------------
 
-const preventDefault = event => event.preventDefault()
-
-class FormModal extends BaseComponent {
-  state = {
-    value: this.props.defaultValue,
-  }
-
-  get value () {
-    return this.state.value
-  }
-
-  render () {
-    const { body, formId } = this.props
-    return (
-      <form id={formId} onSubmit={preventDefault}>
-        {cloneElement(body, {
-          value: this.state.value,
-          onChange: this.linkState('value'),
-        })}
-      </form>
-    )
-  }
-}
+let formModalState
+export const form = ({
+  component,
+  defaultValue,
+  handler = identity,
+  header,
+  render,
+  size,
+}) =>
+  new Promise((resolve, reject) => {
+    formModalState.component = component
+    formModalState.handler = handler
+    formModalState.header = header
+    formModalState.opened = true
+    formModalState.reject = reject
+    formModalState.render = render
+    formModalState.resolve = resolve
+    formModalState.size = size
+    formModalState.value = defaultValue
+    disableShortcuts()
+  })
 
-export const form = ({ body, defaultValue, icon, title, size }) => {
-  const formId = generateRandomId()
-  const buttons = [
-    {
-      btnStyle: 'primary',
-      label: _('formOk'),
-      form: formId,
+const getInitialState = () => ({
+  component: undefined,
+  handler: undefined,
+  header: undefined,
+  isHandlerRunning: false,
+  opened: false,
+  reject: undefined,
+  render: undefined,
+  resolve: undefined,
+  size: undefined,
+  value: undefined,
+})
+export const FormModal = [
+  provideState({
+    initialState: getInitialState,
+    effects: {
+      initialize () {
+        if (formModalState !== undefined) {
+          throw new Error('FormModal is a singleton!')
+        }
+        formModalState = this.state
+      },
+      finalize: () => {
+        formModalState = undefined
+      },
+      onChange: (_, value) => () => ({
+        value: getEventValue(value),
+      }),
+      onCancel () {
+        const { state } = this
+        if (!state.isHandlerRunning) {
+          state.opened = false
+          state.reject()
+        }
+      },
+      async onSubmit ({ close }) {
+        const { state } = this
+        state.isHandlerRunning = true
+
+        let result
+        try {
+          result = await state.handler(state.value)
+        } finally {
+          state.isHandlerRunning = false
+        }
+        state.opened = false
+        state.resolve(result)
+      },
+      reset: () => () => {
```

Comment:
Idem, remove extra function.

Also, you don't need to use an arrow function here.